pkg/sdn/plugin/pod.go

```diff
@@ -0,0 +1,410 @@
+
+	err = containerNs.Do(func(ns.NetNS) error {
+		contVeth, err = netlink.LinkByName(containerIfname)
+		if err != nil {
+			return err
+		}
+		peerIfindex = contVeth.Attrs().ParentIndex
+
+		addrs, err := netlink.AddrList(contVeth, syscall.AF_INET)
+		if err != nil {
+			return fmt.Errorf("failed to get container IP addresses: %v", err)
+		}
+		if len(addrs) == 0 {
+			return fmt.Errorf("container had no addresses")
+		}
+		podIP = addrs[0].IP.String()
+
+		return nil
+	})
+	if err != nil {
+		return nil, "", "", fmt.Errorf("failed to inspect container interface: %v", err)
+	}
+
+	hostVeth, err := netlink.LinkByIndex(peerIfindex)
+	if err != nil {
+		return nil, "", "", fmt.Errorf("failed to get host veth: %v", err)
+	}
+
+	return hostVeth, contVeth.Attrs().HardwareAddr.String(), podIP, nil
+}
+
+func addMacvlan(netns string) error {
+	var defIface netlink.Link
+	var err error
+
+	// Find interface with the default route
+	routes, err := netlink.RouteList(nil, netlink.FAMILY_V4)
+	if err != nil {
+		return fmt.Errorf("failed to read routes: %v", err)
+	}
+
+	for _, r := range routes {
+		if r.Dst == nil {
+			defIface, err = netlink.LinkByIndex(r.LinkIndex)
+			if err != nil {
+				return fmt.Errorf("failed to get default route interface: %v", err)
+			}
+		}
+	}
+	if defIface == nil {
+		return fmt.Errorf("failed to find default route interface")
+	}
+
+	return ns.WithNetNSPath(netns, func(ns.NetNS) error {
+		err := netlink.LinkAdd(&netlink.Macvlan{
+			LinkAttrs: netlink.LinkAttrs{
+				MTU:         defIface.Attrs().MTU,
+				Name:        "macvlan0",
+				ParentIndex: defIface.Attrs().Index,
+			},
+			Mode: netlink.MACVLAN_MODE_PRIVATE,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to create macvlan interface: %v", err)
+		}
+		l, err := netlink.LinkByName("macvlan0")
+		if err != nil {
+			return fmt.Errorf("failed to find macvlan interface: %v", err)
+		}
+		err = netlink.LinkSetUp(l)
+		if err != nil {
+			return fmt.Errorf("failed to set macvlan interface up: %v", err)
+		}
+		return nil
+	})
+}
+
+func getIPAMConfig(clusterNetwork *net.IPNet, localSubnet string) ([]byte, error) {
+	nodeNet, err := types.ParseCIDR(localSubnet)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing node network '%s': %v", localSubnet, err)
+	}
+
+	type hostLocalIPAM struct {
+		Type   string        `json:"type"`
+		Subnet types.IPNet   `json:"subnet"`
+		Routes []types.Route `json:"routes"`
+	}
+
+	type cniNetworkConfig struct {
+		Name string         `json:"name"`
+		Type string         `json:"type"`
+		IPAM *hostLocalIPAM `json:"ipam"`
+	}
+
+	return json.Marshal(&cniNetworkConfig{
+		Name: "openshift-sdn",
+		Type: "openshift-sdn",
+		IPAM: &hostLocalIPAM{
+			Type: "host-local",
+			Subnet: types.IPNet{
+				IP:   nodeNet.IP,
+				Mask: nodeNet.Mask,
+			},
+			Routes: []types.Route{
+				{Dst: net.IPNet{
+					IP:   net.IPv4zero,
+					Mask: net.IPMask(net.IPv4zero),
+				},
+					GW: netutils.GenerateDefaultGateway(nodeNet)},
```

Comment:
This indentation is terrible... gofmt accepts it, but... ugh. There must be some better way.

gofmt also accepts it if you put a newline between "{" and "Dst" and then indents Dst and GW at the same level